lib/internal/modules/esm/hooks.js

```diff
@@ -164,6 +167,16 @@ class Hooks {
    * @returns {any | Promise<any>} User data, ignored unless it's a promise, in which case it will be awaited.
    */
   addCustomLoader(url, exports, data) {
+    const alreadyKnown = ArrayPrototypeSome(ArrayPrototypeMap(['initialize', 'resolve', 'load'], (hookName) => {
+      if (this.#chains[hookName]) {
+        return ArrayPrototypeFilter(this.#chains[hookName], (el) => el.url === url).length === 1;
+      }
+    }), (el) => el);
+
+    if (alreadyKnown) {
+      return undefined;
+    }
+
```

Comment:
considering the `data` could and should be added to the filtering. The reason why I added it was to prevent automatic reregistration due to not only user re-registering unintendedly the same hooks but also because of the automatic spreading that happens in node code. Think of the main usecase: `node --import registerHooks.mjs app.js`:

this gets passed to the hooks thread itself on initialization => two times registered
it also automatically gets passed to any `new Worker()` (in some conditions. Means each such worker will reregister. And now combined with a scenario where the app has short-lived ephemeral threads => longer and longer chain.

Alternative ideas are welcome. But I think it has a usecase that is also beyond what the user app does.

Reply:
Let's remove it and discuss it in a separate PR. In the mean time, it's up to the user to ensure they don't register the same loader twice.

Reply:
@aduh95 I think the issue is that if you start your app with a flag, like `node --import hooks.js app.js`, those flags get automatically inherited by workers unless the user launches them via `new Worker('./worker.js', { execArgv: [] }`. So the default case of `new Worker('./worker.js')` would mean that `hooks.js` gets run again, and those hooks re-registered, for every worker thread that the user creates.

I think the naïve user using a library like `node --import tsx app.js` won’t have any idea that they need to pass `execArgv: []` to avoid `tsx` getting registered a second time; nor should they need to know that. Perhaps `tsx` can somehow be smart enough to know that it’s already registered and therefore avoid calling `register` a second time; is that what you’re proposing?

Reply:
I added logic to consider different `data` with same `url` to be its own loader hook in the chain. Tests are still missing, ongoing work.